nixos/modules/services/security/vault.nix

```
@@ -6,41 +6,31 @@ let
   cfg = config.services.vault;
   opt = options.services.vault;
 
-  configFile = pkgs.writeText "vault.hcl" ''
-    # vault in dev mode will refuse to start if its configuration sets listener
-    ${lib.optionalString (!cfg.dev) ''
-    listener "tcp" {
-      address = "${cfg.address}"
-      ${if (cfg.tlsCertFile == null || cfg.tlsKeyFile == null) then ''
-          tls_disable = "true"
-        '' else ''
-          tls_cert_file = "${cfg.tlsCertFile}"
-          tls_key_file = "${cfg.tlsKeyFile}"
-        ''}
-      ${cfg.listenerExtraConfig}
-    }
-    ''}
-    storage "${cfg.storageBackend}" {
-      ${optionalString (cfg.storagePath   != null) ''path = "${cfg.storagePath}"''}
-      ${optionalString (cfg.storageConfig != null) cfg.storageConfig}
-    }
-    ${optionalString (cfg.telemetryConfig != "") ''
-        telemetry {
-          ${cfg.telemetryConfig}
-        }
-      ''}
-    ${cfg.extraConfig}
-  '';
+  format = pkgs.formats.json {};
+  configFile = format.generate "vault.hcl.json" (filterAttrsRecursive (_: v: v != null) cfg.settings);
```

Filtering out `null` means that settings can't be "nulled out" if they have a non-empty default value in Vault.
I don't know whether this is the case, or will be, but perhaps it could be filtered based on whether the option is defined.
This additionally makes the generated documentation a bit more accurate.

```suggestion
  filteredConfig = filterAttrsRecursive (_name: optOrTree: isOption optOrTree -> optOrTree.isDefined) opt.settings;
  configFile = format.generate "vault.hcl.json" filteredConfig);
```

I haven't tested this, so it probably needs a bit of tweaking here or there, or maybe it's just not feasible.

Note the reference to `opt` instead of `cfg` and the logical implication operator `->`.

With this in place, you can and should remove the `nullOr`s and the `default`s, as well as the `submodule`s, so that `optOrTree` is actually a tree in those nodes.

---

> Filtering out `null` means that settings can't be "nulled out" if they have a non-empty default value in Vault.

right, this is an issue you've discussed in the past. maybe this has something to do with a `record` type you played around with at one point? my memory is a bit foggy...

> I don't know whether this is the case, or will be
> I haven't tested this, so it probably needs a bit of tweaking here or there, or maybe it's just not feasible.

i'm not sure this PR is the place for that type of experimentation - specifically i'm here to solve some real defects in this module (lack of flexibility, can't disable tcp in favour of unix sockets for example) in the standard and RFC accepted way

@roberth **is this answer acceptable as a resolution to this comment?**

> I haven't tested this, so it probably needs a bit of tweaking here or there, or maybe it's just not feasible.
> 
> Note the reference to `opt` instead of `cfg` and the logical implication operator `->`.
> 
> With this in place, you can and should remove the `nullOr`s and the `default`s, as well as the `submodule`s, so that `optOrTree` is actually a tree in those nodes.

without `submodule`s i don't think we get `freeformType`, do we? that is pretty important here
what you're saying is very interesting to me and i would love to hear more about this... please let me know if you have (or have seen) any examples relating to working with modules this way

---

i ended up doing some experimenting with this because your point remains valid and very relevant to nixos modules in general

i found that `isDefined` describes whether a `default` value was set, not whether any value has been set - at least in the case of `submodule` and `getSubOptions`, which appeared to be required here

---

@infinisil as the author of RFC42 are you able to add anything to the conversation? specifically, in the scenario that we have a bunch of freeform options in a `settings` option is there a way to avoid using many `nullOr` types and then filtering them out? using a combination of `isOption` and `isDefined` doesn't seem to work as @roberth suggested it might

any and all feedback is appreciated 🙇‍♂️

---

This is unfortunately a bit of a known hole with RFC42, I believe https://github.com/NixOS/nixpkgs/pull/334680 set out to address it (cc @MattSturgeon)

But for this specific PR, unless there's actually a dire need to be able to set `null` to override vault, I think the current code is fine.

---

thanks!

---

Yes, #333799 and #334680 are both intended to provide a way to have "optional" sub-options, though they do both approach it in a totally different way. I haven't had time recently to continue work on #334680 but I may be able to soon.

I've also been considering other things that may be useful;
- A new `mkOption` argument that allows marking options as optional (similar to #333799, but on a per-option basis. Would likely have the same performance concerns.)
- An `isDefinedExplicitly` option attr, true when `isDefined`, but without using the `default`.


If you're dealing with one or two well-known optional sub-options, you may be able to work-around the issue by
1. Remove the sub-option's `default`
2. Add an `apply` function to the submodule option.
   This apply-function can filter out the sub-option values if the corresponding sub-option is undefined.

If the submodule option itself causes `getSubOptions` to always return options with defaults (I can't recall if it does), you may be able to work around this further by doing some work _within_ the submodule to produce a list of attrs to remove, and then in the option's `apply` function, you could remove the listed attrs, along with the list itself... Not ideal!

---

@MattSturgeon 

> * An `isDefinedExplicitly` option attr, true when `isDefined`, but without using the `default`.

This can also be achieved with a function from option to bool, by looking at the `highestPrio` attribute. This is more efficient, because the large amount of options that are in memory would each need an extra field. Not a big deal normally - just not preferable, because it bloats the already somewhat expensive module system.



> 2\. Add an `apply` function to the submodule option.

This is more strict than necessary, leading to infinite recursions when a setting is defined in terms of another setting in a way that would be valid without the `apply` (and using `lazyAttrsOf` instead of `attrsOf`, if applicable)

I'd recommend to do it right before generating the config, as shown in my suggestion above. It is almost equivalent, but more robust that way.

---

Checking `isDefined` definitely makes sense when an option is deprecated, e.g. via `mkRenameOptionModule` or `mkRemovedOptionModule`.

For rfc42 it _could_ make sense to only include defined options in the generated/formatted output (alongside freeform definitions), but this would be better handled internally (e.g. #333799 or #334680).

I guess the issue you're hinting at is that once there's a definition somewhere, you can't use an override to _un_define it. However doesn't the same argument already apply to freeform definitions?

Regardless, I think I've sufficiently derailed this PR's discussions, so I'll try and refrain from further off-topic discussion 😀

---

OT:

> Tbh I think using `isDefined` is an anti-pattern it general, because once an option is defined, it can never be overridden to be undefined again.

Most configuration systems behave this way, and overriding is itself an antipattern, so this appears almost meaningless to me.

Hot take: if unset-ness matters, it should have an `enable` flag.